-code-checks

ERROR: Build errors found during clang-tidy run.
```
L1Trigger/CSCTriggerPrimitives/interface/CSCPatternBank.h:20:58: error: no member named 'ALCT_PATTERN_WIDTH' in 'CSCConstants'; did you mean 'CLCT_PATTERN_WIDTH'? [clang-diagnostic-error]
  static const int alct_keywire_offset_[2][CSCConstants::ALCT_PATTERN_WIDTH];
                                                         ^~~~~~~~~~~~~~~~~~
                                                         CLCT_PATTERN_WIDTH
--
L1Trigger/CSCTriggerPrimitives/src/CSCPatternBank.cc:3:65: error: no member named 'ALCT_PATTERN_WIDTH' in 'CSCConstants'; did you mean 'CLCT_PATTERN_WIDTH'? [clang-diagnostic-error]
const int CSCPatternBank::alct_keywire_offset_[2][CSCConstants::ALCT_PATTERN_WIDTH] = {
                                                                ^~~~~~~~~~~~~~~~~~
                                                                CLCT_PATTERN_WIDTH
--
L1Trigger/CSCTriggerPrimitives/interface/CSCPatternBank.h:20:58: error: no member named 'ALCT_PATTERN_WIDTH' in 'CSCConstants'; did you mean 'CLCT_PATTERN_WIDTH'? [clang-diagnostic-error]
  static const int alct_keywire_offset_[2][CSCConstants::ALCT_PATTERN_WIDTH];
                                                         ^~~~~~~~~~~~~~~~~~
                                                         CLCT_PATTERN_WIDTH
--
L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc:566:53: error: no member named 'ALCT_PATTERN_WIDTH' in 'CSCConstants'; did you mean 'CLCT_PATTERN_WIDTH'? [clang-diagnostic-error]
        for (int i_wire = 0; i_wire < CSCConstants::ALCT_PATTERN_WIDTH; i_wire++) {
                                                    ^~~~~~~~~~~~~~~~~~
                                                    CLCT_PATTERN_WIDTH
--
L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc:632:30: error: no member named 'ALCT_PATTERN_WIDTH' in 'CSCConstants'; did you mean 'CLCT_PATTERN_WIDTH'? [clang-diagnostic-error]
      p.resize(CSCConstants::ALCT_PATTERN_WIDTH, INVALID_WIRE);
                             ^~~~~~~~~~~~~~~~~~
                             CLCT_PATTERN_WIDTH
--
L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc:640:51: error: no member named 'ALCT_PATTERN_WIDTH' in 'CSCConstants'; did you mean 'CLCT_PATTERN_WIDTH'? [clang-diagnostic-error]
      for (int i_wire = 0; i_wire < CSCConstants::ALCT_PATTERN_WIDTH; i_wire++) {
                                                  ^~~~~~~~~~~~~~~~~~
                                                  CLCT_PATTERN_WIDTH
--
L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc:1361:51: error: no member named 'ALCT_PATTERN_WIDTH' in 'CSCConstants'; did you mean 'CLCT_PATTERN_WIDTH'? [clang-diagnostic-error]
      for (int i_wire = 0; i_wire < CSCConstants::ALCT_PATTERN_WIDTH; i_wire++) {
                                                  ^~~~~~~~~~~~~~~~~~
                                                  CLCT_PATTERN_WIDTH
--
L1Trigger/CSCTriggerPrimitives/interface/CSCPatternBank.h:20:58: error: no member named 'ALCT_PATTERN_WIDTH' in 'CSCConstants'; did you mean 'CLCT_PATTERN_WIDTH'? [clang-diagnostic-error]
  static const int alct_keywire_offset_[2][CSCConstants::ALCT_PATTERN_WIDTH];
                                                         ^~~~~~~~~~~~~~~~~~
                                                         CLCT_PATTERN_WIDTH
--
L1Trigger/CSCTriggerPrimitives/plugins/CSCAnodeLCTAnalyzer.cc:99:51: error: no member named 'ALCT_PATTERN_WIDTH' in 'CSCConstants'; did you mean 'CLCT_PATTERN_WIDTH'? [clang-diagnostic-error]
      for (int i_wire = 0; i_wire < CSCConstants::ALCT_PATTERN_WIDTH; ++i_wire) {
                                                  ^~~~~~~~~~~~~~~~~~
                                                  CLCT_PATTERN_WIDTH
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

--- a/L1Trigger/CSCTriggerPrimitives/interface/CSCAnodeLCTProcessor.h
+++ b/L1Trigger/CSCTriggerPrimitives/interface/CSCAnodeLCTProcessor.h
@@ -198,7 +198,11 @@ protected:
   // enum used in the wire hit assignment
   enum ALCT_WireInfo { INVALID_WIRE = 65535 };
 
+  // remove the invalid wires from the container
   void cleanWireContainer(CSCALCTDigi::WireContainer& wireHits) const;
+
+  //  set the wire hit container
+  void setWireContainer(CSCALCTDigi&, CSCALCTDigi::WireContainer& wireHits) const;
 
   /* This function looks for LCTs on the previous and next wires.  If one
      exists and it has a better quality and a bx_time up to 4 clocks earlier

--- a/L1Trigger/CSCTriggerPrimitives/interface/CSCAnodeLCTProcessor.h
+++ b/L1Trigger/CSCTriggerPrimitives/interface/CSCAnodeLCTProcessor.h
@@ -192,7 +192,13 @@ protected:
   /* See if there is a pattern that satisfies nplanes_hit_pattern number of
      layers hit for either the accelerator or collision patterns.  Use
      the pattern with the best quality. */
-  bool patternDetection(const int key_wire);
+  bool patternDetection(const int key_wire,
+                        std::map<int, std::map<int, CSCALCTDigi::WireContainer> >& hits_in_patterns);
+
+  // enum used in the wire hit assignment
+  enum ALCT_WireInfo { INVALID_WIRE = 65535 };
+
+  void cleanWireContainer(CSCALCTDigi::WireContainer& wireHits) const;
 
   /* This function looks for LCTs on the previous and next wires.  If one
      exists and it has a better quality and a bx_time up to 4 clocks earlier

--- a/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc
+++ b/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc
@@ -295,6 +295,11 @@ void CSCAnodeLCTProcessor::run(const std::vector<int> wire[CSCConstants::NUM_LAY
   // Take the best MAX_CLCTS_PER_PROCESSOR candidates per bx.
   int ALCTIndex_[CSCConstants::MAX_ALCT_TBINS] = {};
 
+  // define a new pattern map
+  // for each key half strip, and for each pattern, store the 2D collection of fired comparator digis
+  std::map<int, std::map<int, CSCCLCTDigi::ComparatorContainer>> hits_in_patterns;
+  hits_in_patterns.clear();
+
   // Only do the rest of the processing if chamber is not empty.
   // Stop drift_delay bx's short of fifo_tbins since at later bx's we will
   // not have a full set of hits to start pattern search anyway.
@@ -307,7 +312,7 @@ void CSCAnodeLCTProcessor::run(const std::vector<int> wire[CSCConstants::NUM_LAY
         if (preTrigger(i_wire, start_bx)) {
           if (infoV > 2)
             showPatterns(i_wire);
-          if (patternDetection(i_wire)) {
+          if (patternDetection(i_wire, hits_in_patterns)) {
             trigger = true;
             int ghost_cleared[2] = {0, 0};
             ghostCancellationLogicOneWire(i_wire, ghost_cleared);
@@ -320,7 +325,17 @@ void CSCAnodeLCTProcessor::run(const std::vector<int> wire[CSCConstants::NUM_LAY
             //acceleration mode
             if (quality[i_wire][0] > 0 and bx < CSCConstants::MAX_ALCT_TBINS) {
               int valid = (ghost_cleared[0] == 0) ? 1 : 0;  //cancelled, valid=0, otherwise it is 1
-              const auto& newALCT(CSCALCTDigi(valid, quality[i_wire][0], 1, 0, i_wire, bx));
+              CSCALCTDigi newALCT(valid, quality[i_wire][0], 1, 0, i_wire, bx);
+
+              // get the comparator hits for this pattern
+              auto wireHits = hits_in_patterns[i_wire][0];
+
+              // purge the wire digi collection
+              cleanWireContainer(wireHits);
+
+              // set the hit collection
+              newALCT.setHits(wireHits);
+
               lct_list.push_back(newALCT);
               ALCTContainer_[bx][ALCTIndex_[bx]] = newALCT;
               ALCTIndex_[bx]++;
@@ -331,7 +346,18 @@ void CSCAnodeLCTProcessor::run(const std::vector<int> wire[CSCConstants::NUM_LAY
             //collision mode
             if (quality[i_wire][1] > 0 and bx < CSCConstants::MAX_ALCT_TBINS) {
               int valid = (ghost_cleared[1] == 0) ? 1 : 0;  //cancelled, valid=0, otherwise it is 1
-              const auto& newALCT(CSCALCTDigi(valid, quality[i_wire][1], 0, quality[i_wire][2], i_wire, bx));
+
+              CSCALCTDigi newALCT(valid, quality[i_wire][1], 0, quality[i_wire][2], i_wire, bx);
+
+              // get the comparator hits for this pattern
+              auto wireHits = hits_in_patterns[i_wire][1];
+
+              // purge the wire digi collection
+              cleanWireContainer(wireHits);
+
+              // set the hit collection
+              newALCT.setHits(wireHits);
+
               lct_list.push_back(newALCT);
               ALCTContainer_[bx][ALCTIndex_[bx]] = newALCT;
               ALCTIndex_[bx]++;
@@ -589,7 +615,8 @@ bool CSCAnodeLCTProcessor::preTrigger(const int key_wire, const int start_bx) {
   return false;
 }
 
-bool CSCAnodeLCTProcessor::patternDetection(const int key_wire) {
+bool CSCAnodeLCTProcessor::patternDetection(
+    const int key_wire, std::map<int, std::map<int, CSCALCTDigi::WireContainer>>& hits_in_patterns) {
   bool trigger = false;
   bool hit_layer[CSCConstants::NUM_LAYERS];
   unsigned int temp_quality;
@@ -609,6 +636,13 @@ bool CSCAnodeLCTProcessor::patternDetection(const int key_wire) {
     for (int i_layer = 0; i_layer < CSCConstants::NUM_LAYERS; i_layer++)
       hit_layer[i_layer] = false;
 
+    // clear a single pattern!
+    CSCALCTDigi::WireContainer hits_single_pattern;
+    hits_single_pattern.resize(CSCConstants::NUM_LAYERS);
+    for (auto& p : hits_single_pattern) {
+      p.resize(CSCConstants::ALCT_PATTERN_WIDTH, INVALID_WIRE);
+    }
+
     double num_pattern_hits = 0., times_sum = 0.;
     std::multiset<int> mset_for_median;
     mset_for_median.clear();
@@ -623,6 +657,9 @@ bool CSCAnodeLCTProcessor::patternDetection(const int key_wire) {
             // Wait a drift_delay time later and look for layers hit in
             // the pattern.
             if (((pulse[i_layer][this_wire] >> (first_bx[key_wire] + drift_delay)) & 1) == 1) {
+              // store hits in the temporary pattern vector
+              hits_single_pattern[i_layer][i_wire] = this_wire;
+
               // If layer has never had a hit before, then increment number
               // of layer hits.
               if (!hit_layer[i_layer]) {
@@ -683,8 +720,10 @@ bool CSCAnodeLCTProcessor::patternDetection(const int key_wire) {
 #endif
     }
 
+    // save the pattern information when a trigger was formed!
     if (temp_quality >= pattern_thresh[i_pattern]) {
       trigger = true;
+      hits_in_patterns[key_wire][i_pattern] = hits_single_pattern;
 
       // Quality definition changed on 22 June 2007: it no longer depends
       // on pattern_thresh.
@@ -1356,4 +1395,12 @@ int CSCAnodeLCTProcessor::getTempALCTQuality(int temp_quality) const {
     Q = 0;  // quality code 0 is valid!
 
   return Q;
+}
+
+void CSCAnodeLCTProcessor::cleanWireContainer(CSCALCTDigi::WireContainer& wireHits) const {
+  for (auto& p : wireHits) {
+    p.erase(
+        std::remove_if(p.begin(), p.end(), [](unsigned i) -> bool { return i == CSCAnodeLCTProcessor::INVALID_WIRE; }),
+        p.end());
+  }
 }

--- a/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc
+++ b/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc
@@ -327,14 +327,8 @@ void CSCAnodeLCTProcessor::run(const std::vector<int> wire[CSCConstants::NUM_LAY
               int valid = (ghost_cleared[0] == 0) ? 1 : 0;  //cancelled, valid=0, otherwise it is 1
               CSCALCTDigi newALCT(valid, quality[i_wire][0], 1, 0, i_wire, bx);
 
-              // get the comparator hits for this pattern
-              auto wireHits = hits_in_patterns[i_wire][0];
-
-              // purge the wire digi collection
-              cleanWireContainer(wireHits);
-
-              // set the hit collection
-              newALCT.setHits(wireHits);
+              // set the wire digis for this pattern
+              setWireContainer(newALCT, hits_in_patterns[i_wire][0]);
 
               lct_list.push_back(newALCT);
               ALCTContainer_[bx][ALCTIndex_[bx]] = newALCT;
@@ -349,14 +343,8 @@ void CSCAnodeLCTProcessor::run(const std::vector<int> wire[CSCConstants::NUM_LAY
 
               CSCALCTDigi newALCT(valid, quality[i_wire][1], 0, quality[i_wire][2], i_wire, bx);
 
-              // get the comparator hits for this pattern
-              auto wireHits = hits_in_patterns[i_wire][1];
-
-              // purge the wire digi collection
-              cleanWireContainer(wireHits);
-
-              // set the hit collection
-              newALCT.setHits(wireHits);
+              // set the wire digis for this pattern
+              setWireContainer(newALCT, hits_in_patterns[i_wire][1]);
 
               lct_list.push_back(newALCT);
               ALCTContainer_[bx][ALCTIndex_[bx]] = newALCT;
@@ -638,6 +626,7 @@ bool CSCAnodeLCTProcessor::patternDetection(
 
     // clear a single pattern!
     CSCALCTDigi::WireContainer hits_single_pattern;
+    hits_single_pattern.clear();
     hits_single_pattern.resize(CSCConstants::NUM_LAYERS);
     for (auto& p : hits_single_pattern) {
       p.resize(CSCConstants::ALCT_PATTERN_WIDTH, INVALID_WIRE);
@@ -1403,4 +1392,12 @@ void CSCAnodeLCTProcessor::cleanWireContainer(CSCALCTDigi::WireContainer& wireHi
         std::remove_if(p.begin(), p.end(), [](unsigned i) -> bool { return i == CSCAnodeLCTProcessor::INVALID_WIRE; }),
         p.end());
   }
+}
+
+void CSCAnodeLCTProcessor::setWireContainer(CSCALCTDigi& alct, CSCALCTDigi::WireContainer& wireHits) const {
+  // clean the wire digi container
+  cleanWireContainer(wireHits);
+
+  // set the hit container
+  alct.setHits(wireHits);
 }

